python/kserve/kserve/protocol/rest/openai/openai_model.py

```diff
@@ -0,0 +1,288 @@
+# Copyright 2023 The KServe Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import AsyncIterator, Callable, Iterable, Union, cast
+
+from openai.types import Completion, CompletionChoice, CompletionCreateParams
+from openai.types.chat import ChatCompletion, ChatCompletionChunk
+from openai.types.chat import ChatCompletionMessage as BaseChatCompletionMessage
+from openai.types.chat import ChatCompletionMessageParam
+from openai.types.chat import CompletionCreateParams as ChatCompletionCreateParams
+from openai.types.chat.chat_completion import Choice, ChoiceLogprobs
+from openai.types.chat.chat_completion_chunk import Choice as ChunkChoice
+from openai.types.chat.chat_completion_chunk import ChoiceDelta
+from openai.types.chat.chat_completion_chunk import (
+    ChoiceLogprobs as ChunkChoiceLogprobs,
+)
+from openai.types.chat.chat_completion_token_logprob import (
+    ChatCompletionTokenLogprob,
+    TopLogprob,
+)
+from openai.types.completion_choice import Logprobs
+from openai.types.completion_create_params import (
+    CompletionCreateParamsNonStreaming,
+    CompletionCreateParamsStreaming,
+)
+from pydantic import BaseModel
+
+from ....errors import InvalidInput
+
+
+class ChatPrompt(BaseModel):
+    response_role: str = "assistant"
+    prompt: str
+
+
+class ChatCompletionMessage(BaseChatCompletionMessage):
+    role: str
+
+
+class OpenAIModel(ABC):
+    """
+    An abstract model with methods for implementing OpenAI's completions (v1/completions)
+    and chat completions (v1/chat/completions) endpoints.
+
+    Users should extend this model and implement the abstract methods in order to expose
+    these endpoints.
+    """
+
+    @abstractmethod
+    async def create_completion(
+        self, params: CompletionCreateParams
+    ) -> Union[Completion, AsyncIterator[Completion]]:
+        pass
+
+    @abstractmethod
+    async def create_chat_completion(
+        self, params: ChatCompletionCreateParams
+    ) -> Union[ChatCompletion, AsyncIterator[ChatCompletionChunk]]:
+        pass
+
```

Comment:
The [model implementation we used in V1 endpoints](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/model.py#L82) was baked in with the `preprocess`/`predict`/`postprocess` api. It looks like we are moving away from this. Was this a conscious choice?

This also means users are expected to implement the [predict logic](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/model.py#L253) that used to come out of the box. This includes protocol selection, forming the target host, handling headers, errors etc. Do I understand this correctly? Or does the openai sdk handle this for us?

Reply:
It's a conscious choice as we're strictly following OpenAI API which doesn't have `preprocess`/`predict`/`postprocess`, only `create_chat_completion`. Users will have to implement the prediction logic but we've provided the adapter to convert between `Completion` and `ChatCompletion` types. If users are using `huggingfaceserver` however then `create_chat_completion` is provided for free.

Reply:
After discussing with @yuzisun, we'll add `preprocess()` and `postprocess()` to the `OpenAIModel` abstract class. So the workflow should look like: `preprocess()` -> `create_chat_completion()` -> `postprocess()`.